[REFACTOR]: Loads of things

SeedyROM opened the pull request:
This allows for error propagation and not a unhandled panic!

- [x] Remove unwanted expects and nested closures.
- [x] Fix left over `Box<dyn error>`
- [x] Try to join the tasks at the end and propagate the error.
- [x] Stop dead lock for block_stream_tx. 

Most Likely Fixes: #56 and will help with other error handling issues in the future.

SeedyROM committed:
:hammer: Remove unwarrented panics and extra closures from system/mod.rs

SeedyROM committed:
:hammer: Actually try the join handles and check the result

SeedyROM committed:
:hammer: Fix tracing info to be an error log

SeedyROM committed:
:hammer: Fix tracing info to be an error log

SeedyROM committed:
:hammer: Possibly handle block polling crash

SeedyROM committed:
:hammer: Add notes on getting retry working

TrevorJTClarke commented on croncat/src/system/mod.rs:
```diff
@@ -7,6 +7,11 @@ use std::time::Duration;
 use cw_croncat_core::types::AgentStatus;
 use tokio::sync::Mutex;
+use tokio_retry::{
+    strategy::{jitter, ExponentialBackoff},
```
Getting some over-caffeinated vibes here ;)

TrevorJTClarke reviewed (commented):
This all looks great!
I need to run it, but am AFKish for truly testing. 
So it looks like its all good, just wanna give it a go

SeedyROM committed:
:hammer: Organize deps for croncat crate

SeedyROM committed:
:hammer: Add set overflow for broadcast block channel, fix error propagation for ws.rs

SeedyROM committed:
Merge branch 'main' into refactor/system

Buckram123 commented on croncat/src/channels.rs:
```diff
@@ -18,7 +18,11 @@ pub type BlockStreamRx = async_broadcast::Receiver<tendermint::Block>;
+    let mut channel = async_broadcast::broadcast(channel_size);
+    // TODO: THIS IS A HACK. Something is not flushing the channel properly.
+    // For now all our tasks should be received fine.
+    channel.0.set_overflow(true);
```
Great solution!

Buckram123 replied:
BTW don't think it is a `hack`, don't think we want to use any RAM to process old blocks and we don't want it to crash in case of queue overflow or even worse OOM

SeedyROM replied:
It is kind of a hack, what was happening is that the broadcast channel would fill up even with all the listeners using the messages. This was causing a deadlock and it's what was making the agent "crash" when the channel buffer was full.

d3v3us replied:
When overflow is set to true, it will flush by removing old messages right?

SeedyROM replied:
Yup.

Buckram123 commented on croncat/src/system/mod.rs:
```diff
@@ -32,77 +37,83 @@ pub async fn run(
+            block_status,
+        ))
+    } else {
+        tokio::task::spawn(async move { Ok(()) })
```
Does it move anything inside?
```suggestion
        tokio::task::spawn(async { Ok(()) })
```

SeedyROM replied:
Good call, let me fix that!

Buckram123 commented on croncat/src/system/mod.rs:
```diff
@@ -32,77 +37,83 @@ pub async fn run(
+    // This should be repeated for all the other tasks probably?
+    //
+    let retry_polling_handle = tokio::task::spawn(async move {
+        Retry::spawn(retry_strategy.clone(), || async {
```
Thanks for showing this cool crate

d3v3us replied:
tokio_retry?

Buckram123 replied:
> tokio_retry?

Yeah, I'm kinda new in `async`)

Buckram123 commented on croncat/src/streams/ws.rs:
```diff
@@ -21,7 +23,7 @@ pub async fn stream_blocks_loop(
     block_stream_tx: BlockStreamTx,
     mut shutdown_rx: ShutdownRx,
-) -> Result<(), Box<dyn std::error::Error>> {
+) -> Result<(), Report> {
```
🧨

Buckram123 reviewed (approved):
Nice fixes and error handlings!

SeedyROM committed:
Update croncat/src/system/mod.rs

Co-authored-by: Mykhailo Donchenko <91957742+Buckram123@users.noreply.github.com>

SeedyROM committed:
:tada: Properly handle retry and failure for polling loop and websocket stream

SeedyROM committed:
Merge branch 'refactor/system' of github.com:SeedyROM/croncat-rs into refactor/system

d3v3us commented on croncat/src/system/mod.rs:
```diff
@@ -32,100 +38,124 @@ pub async fn run(
-            .expect("Failed to stream blocks")
+
+    // Generic retry strategy
+    let retry_strategy = ExponentialBackoff::from_millis(10).map(jitter).take(4);
```
Exponential Backoff, ahh, awesome!)

SeedyROM committed:
Merge branch 'main' into refactor/system

SeedyROM committed:
:hammer: Print full backtrace for easier debugging workflow

SeedyROM committed:
:hammer: Remove sensitive info from Debug printing, implement Debug manually

SeedyROM committed:
:book: Refactor main.rs to have commands run in their own function, makes handling command errors and logging easier

SeedyROM committed:
:hammer: Fix typo for derivation

SeedyROM committed:
:hammer: Remove full backtrace, it gets too insane at nested levels

SeedyROM committed:
:hammer: Allow for environment variable override for contract address config

SeedyROM committed:
:hammer: Update doc for null contract_address

SeedyROM committed:
:hammer: Oops remove debug print

SeedyROM committed:
:notes: Print main croncatd error handler to stderr

Buckram123 commented on croncat/src/grpc.rs:
```diff
@@ -63,15 +63,15 @@ impl GrpcSigner {
             .client
             .query_client
-            .query_contract(&self.client.cfg.contract_address, msg)
+            .query_contract(&self.client.cfg.contract_address.to_owned().unwrap(), msg)
```
How would you feel to make new struct like `VerifiedChainConfig`, so we have less to `unwrap()`. 

And I think here is redundant clone?
```suggestion
            .query_contract(self.client.cfg.contract_address.as_ref().unwrap(), msg)
```

SeedyROM replied:
Missed that those were being passed by reference! Fixing that now. 

Also that's a possibility with the struct, let me think about that.

Buckram123 commented on croncat/src/logging.rs:
```diff
@@ -12,7 +12,8 @@ pub use tracing::{debug, error, info, warn};
 pub fn setup() -> Result<(), Report> {
-    // Get / set backtrace
+    // Set 1 backtrace if debug.
+    #[cfg(debug_assertions)]
```
Nice

Buckram123 commented on croncat/src/config/mod.rs:
```diff
@@ -41,7 +41,20 @@ impl ChainConfig {
+        let mut config = settings.try_deserialize::<ChainConfig>()?;
+
+        // Override config contract address if env var is set
+        if std::env::var("CRONCAT_CONTRACT_ADDRESS").is_ok() {
+            let contract_address = std::env::var("CRONCAT_CONTRACT_ADDRESS").unwrap();
+            config.contract_address = Some(contract_address);
+        }
+
+        if config.contract_address.is_none() {
```
Thanks, that's awesome we can override contract addr with env var now
Small nit:
```suggestion
        if let Ok(addr) = std::env::var("CRONCAT_CONTRACT_ADDRESS") {
            config.contract_address = Some(addr);
        } else if config.contract_address.is_none() {
```

SeedyROM replied:
:+1:

Buckram123 commented on croncat/src/store/agent.rs:
```diff
@@ -183,10 +199,23 @@ impl LocalAgentStorage {
+        let found = self.data.get(account_id);
+
+        if found.is_some() {
+            info!("Found agent: {:#?}", found.unwrap());
```
Sorry for being pendantic here, I'm pretty sure it's optimized, so it doesn't look twice inside `Option` here, but anyway:
```suggestion
        if let Some(entry) = found {
            info!("Found agent: {:#?}", entry);
```

SeedyROM replied:
Not too pendantic at all! Much better choice of syntax.

Buckram123 commented on croncat/src/streams/polling.rs:
```diff
@@ -15,45 +18,62 @@ use crate::logging::info;
-        _ = polling_loop_handle => {}
-        _ = shutdown_rx.recv() => {}
+        res = polling_loop_handle => {
+            match res? {
+                Ok(_) => {
+
+                    return Ok(())
+                }
+                Err(err) => {
+                    error!("Block polling loop failed: {}", err);
+                    return Err(err.into());
+                }
+            }
+        }
+        _ = shutdown_rx.recv() => {
+            return Ok(())
```
```suggestion
            res?.map_err(|err| {
                error!("Block polling loop failed: {}", err);
                err
            })
        }
        _ = shutdown_rx.recv() => Ok(())
```

SeedyROM replied:
So the reason I kept the full pattern and didn't use .map_err is because we might have work to do on shutdown/closure. In the ws stream module you can see the use case of that happening. For now it's a little unwieldy but I think I'd rather have the pattern be there in stone.

SeedyROM replied:
Actually I think there is a way to simplify it!

Buckram123 replied:
I see, thanks, thought they shutdown on `drop`

Buckram123 commented on croncat/src/utils.rs:
```diff
@@ -7,10 +7,13 @@ use std::sync::{
 
 pub const DEFAULT_AGENT_ID: &str = "agent";
-pub const DERVIATION_PATH: &str = "m/44'/118'/0'/0/0";
+pub const DERIVATION_PATH: &str = "m/44'/118'/0'/0/0";
```
Lol, sharp eye

Buckram123 commented on croncat/src/utils.rs:
```diff
@@ -49,3 +52,14 @@ pub fn sum_num_tasks(tasks: &AgentTaskResponse) -> u64 {
+///
+/// Flatten join handle results.
+///
+pub async fn flatten_join<T>(handle: JoinHandle<Result<T, Report>>) -> Result<T, Report> {
+    match handle.await {
+        Ok(Ok(result)) => Ok(result),
+        Ok(Err(err)) => Err(err),
+        Err(err) => Err(err.into()),
+    }
+}
```
That's so cool

Buckram123 commented on croncatd/src/main.rs:
```diff
@@ -25,15 +26,15 @@ mod opts;
-        .map_err(|e| {
-            println!("{}", e);
+        .map_err(|err| {
+            eprintln!("{}", err);
```
nice!

Buckram123 commented on croncat/src/system/mod.rs:
```diff
@@ -32,100 +38,124 @@ pub async fn run(
-        rules_runner_handle,
-        polling_handle
+    // Try to join all the system tasks.
+    let system_status = tokio::try_join!(
+        flatten_join(ctrl_c_handle),
+        flatten_join(account_status_check_handle),
+        flatten_join(retry_block_stream_handle),
+        flatten_join(retry_polling_handle),
+        flatten_join(task_runner_handle),
+        flatten_join(rules_runner_handle),
     );
```
🧠

Buckram123 commented on config.local.yaml:
```diff
@@ -4,6 +4,6 @@ chain_id: "testing"
 grpc_endpoint: "http://localhost:9090/"
 wsrpc_endpoint: "ws://localhost:26657/websocket"
-contract_address: "juno14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9skjuwg8"
+contract_address: null # This can be set via CRONCAT_CONTRACT_ADDRESS, committing this is a pain.
```
Nice comment

TrevorJTClarke replied:
@SeedyROM i agree its a pain for local, but we need to make sure we dont lose the official addresses for each network. Agents shouldn't be responsible for config, we will need to publish versions of the official addresses.

So actually, (probably diff issue), is there a better way to manage each chains contract addresses? 
We will have to maintain at least 1 address per supported network.

SeedyROM replied:
@TrevorJTClarke Right, I agree. I'm allowing for the override manually. If the config has the contract addr and no env var is set it uses the config address.

SeedyROM replied:
Also local environments will always have different contract addresses right? How could you deploy a contract on a new chain at the same addr?

SeedyROM replied:
I'm not sure about another way to manage the chain addresses outside of config. I'm open to suggestions though!

Buckram123 reviewed (approved):
Good job @SeedyROM!
I hope this gets merged ASAP, getting feeling this thing gonna grow indefinitely :DD

SeedyROM committed:
:hammer: Add suggestion from Buckram123

SeedyROM committed:
:hammer: Use enum-display to show CLI commands

Properly show enum variant names for Command in croncatd.
Currently using the enum-display crate which I wrote just for this
project.

d3v3us commented on croncatd/src/opts.rs:
```diff
@@ -19,7 +20,8 @@ pub struct Opts {
 
-#[derive(Debug, StructOpt, Clone)]
+#[derive(Debug, StructOpt, Clone, EnumDisplay)]
+#[enum_display(case = "Kebab")]
```
Awesome solution, which cases it supports beside Kebab?

SeedyROM replied:
Any case from https://docs.rs/convert_case/latest/convert_case/

SeedyROM committed:
:hammer: Audit all uses of unwrap and expect to make sure we always propagate errors up
